lib/timeline-window.js

```diff
@@ -0,0 +1,392 @@
+ *
+ * @param {number} size   number of events to try to extend by. If fewer than this
+ *    number are immediately available, then we return immediately rather than
+ *    making an API call.
+ *
+ * @param {boolean} [makeRequest = true]  whether we should make API calls to
+ *    fetch further events if we don't have any.
+ *
+ * @return {module:client.Promise} Resolves to a boolean which is true if more events
+ *    were successfully retrieved.
+ */
+TimelineWindow.prototype.paginate = function(direction, size, makeRequest) {
+    // Either wind back the message cap (if there are enough events in the
+    // timeline to do so), or fire off a pagination request.
+
+    if (makeRequest === undefined) {
+        makeRequest = true;
+    }
+
+    var tl;
+    if (direction == EventTimeline.BACKWARDS) {
+        tl = this._start;
+    } else if (direction == EventTimeline.FORWARDS) {
+        tl = this._end;
+    } else {
+        throw new Error("Invalid direction '" + direction + "'");
+    }
+
+    if (!tl) {
+        debuglog("TimelineWindow: no timeline yet");
+        return q(false);
+    }
+
+    if (tl.pendingPaginate) {
+        return tl.pendingPaginate;
+    }
+
+    // try moving the cap
+    var count = (direction == EventTimeline.BACKWARDS) ?
+        tl.retreat(size) : tl.advance(size);
+
+    if (count) {
+        this._eventCount += count;
+        debuglog("TimelineWindow: increased cap by " + count +
+                 " (now " + this._eventCount + ")");
+        // remove some events from the other end, if necessary
+        if (this._eventCount > this._windowLimit) {
+            this._unpaginate(direction != EventTimeline.BACKWARDS);
+        }
+        return q(true);
+    }
+
+    if (!makeRequest) {
+        return q(false);
+    }
+
+    // try making a pagination request
+    var token = tl.timeline.getPaginationToken(direction);
+    if (!token) {
+        debuglog("TimelineWindow: no token");
+        return q(false);
+    }
+
+    debuglog("TimelineWindow: starting request");
+    var self = this;
+    var prom = this._client.paginateEventTimeline(tl.timeline, {
+        backwards: direction == EventTimeline.BACKWARDS,
+        limit: size
+    }).finally(function() {
+        tl.pendingPaginate = null;
+    }).then(function(r) {
+        debuglog("TimelineWindow: request completed with result " + r);
+        if (!r) {
+            // end of timeline
+            return false;
+        }
+        return self.paginate(direction, size);
+    });
+    tl.pendingPaginate = prom;
+    return prom;
+};
+
+
+/**
+ * Trim the window to the windowlimit
+ *
+ * @param {boolean} startOfTimeline if events should be removed from the start
+ *     of the timeline.
+ *
+ * @private
+ */
+TimelineWindow.prototype._unpaginate = function(startOfTimeline) {
+    var tl = startOfTimeline ? this._start : this._end;
+
+    var delta = this._eventCount - this._windowLimit;
```

Comment:
I would prefer if we hoisted this to be the caller's responsibility because we are actually doing this check in https://github.com/matrix-org/matrix-js-sdk/pull/64/files#diff-e0497ffe12218015b7ef5f92abdc85a2R203 - This also makes the function signature more pleasing: `_unpaginate(num, startOfTimeline)`